Update lucene branch

LoayGhreeb opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link the issue that will be closed, e.g., "Closes #333".
If your PR closes a koppor issue, link it using its URL, e.g., "Closes https://github.com/koppor/jabref/issues/47".
"Closes" is a keyword GitHub uses to link PRs with issues; do not change it.
Don't reference an issue in the PR title because GitHub does not support auto-linking there.
-->

This is a follow-up to https://github.com/JabRef/jabref/pull/8963. We target `main`, because targetting `luceneSearchBackend` leads to much more "unrelated" commits.

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented on src/main/java/org/jabref/logic/search/retrieval/LuceneSearcher.java:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    private static String getFieldContents(StoredFields storedFields, ScoreDoc scoreDoc, String field) {
```
Please try to avoid `String`. You can use `enum` somehow. Maybe, also `.toString()` if you interact with lucene. But before, one should use enum for type safety.

koppor commented on src/main/java/org/jabref/gui/maintable/MainTableDataModel.java:
```diff
@@ -16,73 +14,132 @@
+                    .forEach(group -> group.updateMatches(luceneManager));
+    }
+
+    private void doSearch(Optional<SearchQuery> query) {
```
Please try to avoid to pass `Optional`.

Maybe its ok here, because of the `lastSearchQuery`. But then please document it in the JavaDoc that it is necessary.

In general, it should be avoided. Normal Java style: create two methods: One with the parameter and another one without.

koppor commented:
@LoayGhreeb Can you explain following issue?

```
/home/runner/work/jabref/jabref/src/test/java/org/jabref/logic/search/DatabaseSearcherWithBibFilesTest.java:14: error: package org.jabref.logic.pdf.search does not exist
import org.jabref.logic.pdf.search.PdfIndexer;
                                  ^
```

LoayGhreeb commented:
> Can you explain following issue?
I moved the search classes out of the PDF package and removed the `PdfIndexer` class. However, I have not yet adapted the test classes.

koppor commented:
- [ ] Fix https://github.com/JabRef/jabref-issue-melting-pot/issues/12

koppor commented:
I triggered a merge. Maybe, the tests will work then again? - I also hope that we will see a build at https://builds.jabref.org/pull/11326/.

koppor commented:
This branch is coming from an external repository; therefore, no secrets for build upload are present. Testers need to compile for themselves on the local machine.

Depending on the issues, we either work towards a quick merge into main - or create a branch inside the JabRef main repository (and maybe create a new pull request).

LoayGhreeb commented:
> This branch is coming from an external repository; therefore, no secrets for build upload are present. Testers need to compile for themselves on the local machine.
> 
> Depending on the issues, we either work towards a quick merge into main - or create a branch inside the JabRef main repository (and maybe create a new pull request).

I will create a new branch and a new pull request. This PR should be merged after the search floating mode PR is merged.
